clang/lib/Sema/SemaType.cpp

```diff
@@ -2751,19 +2751,22 @@ static void checkExtParameterInfos(Sema &S, ArrayRef<QualType> paramTypes,
                     llvm::function_ref<SourceLocation(unsigned)> getParamLoc) {
   assert(EPI.ExtParameterInfos && "shouldn't get here without param infos");
 
-  bool hasCheckedSwiftCall = false;
-  auto reportWrongABI = [&](unsigned paramIndex) {
-    S.Diag(getParamLoc(paramIndex), diag::err_swift_param_attr_not_swiftcall)
-        << getParameterABISpelling(EPI.ExtParameterInfos[paramIndex].getABI());
+  bool emittedError = false;
+  auto actualCC = EPI.ExtInfo.getCC();
+  enum class RequiredCC {
+    OnlySwift, SwiftOrSwiftAsync
   };
-  auto checkForSwiftCC = [&](unsigned paramIndex) {
-    // Only do this once.
-    if (hasCheckedSwiftCall) return;
-    hasCheckedSwiftCall = true;
-    if (EPI.ExtInfo.getCC() == CC_Swift) return;
-    reportWrongABI(paramIndex);
+  auto checkCompatible = [&](unsigned paramIndex, RequiredCC required) {
+    bool isCompatible = (required == RequiredCC::OnlySwift)
+      ? (actualCC == CC_Swift)
+      : (actualCC == CC_Swift || actualCC == CC_SwiftAsync);
+    if (isCompatible || emittedError) return;
+    S.Diag(getParamLoc(paramIndex), diag::err_swift_param_attr_not_swiftcall)
+        << getParameterABISpelling(EPI.ExtParameterInfos[paramIndex].getABI())
+        << (required == RequiredCC::OnlySwift
+            ? "swiftcall" : "swiftcall or swiftasynccall");
```

Comment:
You should avoid doing things like this that assume that diagnostics are in English.  Use a switch in the diagnostic text.

Reply:
Replaced with `%select` in diagnostic.